rfd/0167-console-service.md

```diff
@@ -0,0 +1,255 @@
+				return net.Dial("unix", socketAddr)
+			},
+		}
+	}
+}
+
+// While making requests using the client, the users will still require to
+// provide the full URL since the client always validate them.
+res, err := client.Get("http://console-service/debug/pprof/heap")
+```
+</details>
+
+#### Changing log level command
+
+Changes the application log level.
+
+`teleport [-c config-path] console set-log-level [LEVEL]`
+
+|Flag|Description|Default value|
+|----|-----------|-------------|
+|`-c,--config`|Teleport configuration path.|`/etc/teleport.yaml`|
+|`LEVEL`|Log level (case-insensitive). Any of: `TRACE`, `DEBUG`, `INFO`, `WARN`, `ERROR`|Required|
+
+Usage examples:
+```bash
+$ teleport console set-log-level DEBUG
+$ teleport -c /random/teleport.yaml console set-log-level INFO
+```
+
+#### Capture `pprof` profiles command
+
+Export the application profile (`pprof` format).
+
+`teleport [-c config-path] console profile [--tar] [--seconds=] [PROFILES]`
+
+The `PROFILES` values follow the Golang's definition on `runtime.Profile`
+plus the profiles defined by `net/http/pprof`:
+- `allocs`: A sampling of all past memory allocations
+- `block`: Stack traces that led to blocking on synchronization primitives
+- `cmdline`: The command line invocation of the current program
+- `goroutine`: Stack traces of all current goroutines. Use debug=2 as a query parameter to export in the same format as an unrecovered panic.
+- `heap`: A sampling of memory allocations of live objects. You can specify the gc GET parameter to run GC before taking the heap sample.
+- `mutex`: Stack traces of holders of contended mutexes
+- `profile`: CPU profile. You can specify the duration in the seconds GET parameter. After you get the profile file, use the go tool pprof command to investigate the profile.
+- `threadcreate`: Stack traces that led to the creation of new OS threads
+- `trace`: A trace of execution of the current program. You can specify the duration in the seconds GET parameter. After you get the trace file, use the go tool trace command to investigate the trace.
+
+Users can profile can collect multiple profiles on a single command call. The
+profile names need to be comma-separated. The command outputs all selected
+profiles into a single compressed tarball.
+
+Note: `--seconds` argument has the same effect as
+[`seconds` query string](https://pkg.go.dev/net/http/pprof#hdr-Parameters).
+This parameter max value will be set to the HTTP server write timeout value.
+
+|Flag|Description|Default value|
+|----|-----------|-------------|
+|`-c,--config`|Teleport configuration path.|`/etc/teleport.yaml`|
+|`-s,--seconds`|For CPU and trace profiles, profile for the given duration. For other profiles, return a delta profile.|None|
+|`PROFILES`|Comma-separated profile names to be exported. Supported profiles: `allocs`, `block`, `cmdline`, `goroutine`, `heap`, `mutex`, `profile`, `threadcreate`, `trace`|`heap,profile,goroutine`|
```

Comment:
> Parameters can be passed via GET query params:
> debug=N (all profiles): response format: N = 0: binary (default), N > 0: plaintext
gc=N (heap profile): N > 0: run a garbage collection cycle before profiling
seconds=N (allocs, block, goroutine, heap, mutex, threadcreate profiles): return a delta profile
seconds=N (cpu (profile), trace profiles): profile for the given duration

We need a path to capture both `delta` and snapshots, for profiles like `heap`.

Reply:
> We need a way to capture both delta and snapshots

not necessarily in the same command

Reply:
Only `trace` and `profile` require the `seconds` argument not to be empty (default value if empty is `30`). The other supports "snapshot" and "delta" profiles.

That said, we could flex the command to explicitly require the `seconds` argument when any of these (`trace` or `profile`) are required. Any other profile provided will be delta (due to the presence of the `seconds` flag). Users must perform the command multiple times to grab snapshots and delta.

I like this idea because it keeps closer to the `pprof` endpoint behavior and makes profiles consistent when requiring more than one per command execution.

On the UX side, we can clarify this by adding info logs and sending them to `stderr` to avoid breaking the pipe functions.

```bash
# Cases where `profile` is present.
$ teleport console profile -s 10 profile,goroutine > pprof.tar.gz
Collected delta profiles (10 seconds).

# When not present, and the seconds flag is provided.
$ teleport console profile -s 10 goroutine,heap > pprof.tar.gz
Collected delta profiles (10 seconds).

# When not present, and no seconds is provided.
$ teleport console profile -s 10 goroutine,heap > pprof.tar.gz
Collected snapshot profiles.

# Validation error on `trace` and `profile` profiles.
$ teleport console profile trace,profile > pprof.tar.gz
"--seconds/-s" flag is required when collecting `trace` and/or `profile` profiles.
```

What do you think @greedy52?

Reply:
sounds good to me 👍

Reply:
> My 2c: lets make this as easy as possible for users, especially the ones that have no idea what pprof is. I would imagine that in most cases users will not be providing -s and if they do and the profile they requested doesn't support it we should maybe drop it from the internal pprof request entirely. I also don't think that the pprof endpoints require that seconds is provided and will default to 1s.

`heap` for example, a snapshot dump (no `seconds` provided to pprof) vs a 30-seconds delta dump has their own purposes. For average users, I assume they will just run `teleport console profile`.  If we want to keep it simple, we can do what's currently documented to skip `seconds` for `heap`. I would still like some way to capture a heap delta though. I guess we can still do it with curl.

Reply:
Users can still collect both snapshot and delta by making the seconds flag not required and not forcing a default value to it (passing empty value to the `pprof` endpoints). The only thing they need to be aware of is that if providing the flag, it will affect all the selected profiles.